lib/phase/merge.js

```diff
@@ -0,0 +1,113 @@
+const TEMPLATE_NAME_REGEX_WITH_NAMESPACE =
+    require('screwdriver-data-schema').config.regex.FULL_TEMPLATE_NAME_WITH_NAMESPACE;
+
+/**
+ * Fetch pipeline template
+ * @method fetchPipelineTemplate
+ * @method
+ * @param   {String}                          pipelineTemplateName                        Pipeline template name
+ * @param   {PipelineTemplateVersionFactory}  config.pipelineTemplateVersionFactory       PiplineTemplateVersion Factory to get pipeline templates
+ * @param   {PipelineTemplateTagFactory}      config.pipelineTemplateTagFactory           PipelineTemplateTag Factory to get pipeline templates tag
+ * @param   {PipelineTemplateFactory}         config.pipelineTemplateFactory              PipelineTemplate Factory to get pipeline templates
+ * @return  {Object}                                                                      Pipeline Template
+ */
+async function fetchPipelineTemplate(
+    pipelineTemplateName,
+    pipelineTemplateVersionFactory,
+    pipelineTemplateTagFactory,
+    pipelineTemplateFactory
+) {
+    const [, namespace, name, versionOrTag] = TEMPLATE_NAME_REGEX_WITH_NAMESPACE.exec(pipelineTemplateName);
+    const templateTag = await pipelineTemplateTagFactory.get({
+        name,
+        namespace,
+        tag: versionOrTag
+    });
+    const version = templateTag ? templateTag.version : versionOrTag;
+    const pipelineTemplate = await pipelineTemplateVersionFactory.getWithMetadata(
+        {
+            name,
+            namespace,
+            version
+        },
+        pipelineTemplateFactory
+    );
+
+    return pipelineTemplate;
+}
+
+/**
+ * Merge parsed yaml setting with pipeline template setting
+ * Pipeline yaml will take precedence over pipeline template
+ * @method mergeSetting
+ * @method
+ * @param   {Object}        parsedDoc              Parsed screwdriver yaml
+ * @param   {Object}        pipelineTemplate       Pipeline template
+ * @return  {Object}                               Merged yaml settings
+ */
+function mergeSetting(parsedDoc, pipelineTemplate) {
+    const newPipeline = pipelineTemplate.config;
+    const yamlSharedConfig = parsedDoc.shared || {};
+    const templateSharedConfig = newPipeline.shared || {};
+
+    templateSharedConfig.environment = Hoek.merge(templateSharedConfig.environment || {}, {
+        SD_PIPELINE_TEMPLATE_FULLNAME: parsedDoc.template,
+        SD_PIPELINE_TEMPLATE_NAME: pipelineTemplate.name,
+        SD_PIPELINE_TEMPLATE_NAMESPACE: pipelineTemplate.namespace,
+        SD_PIPELINE_TEMPLATE_VERSION: pipelineTemplate.version
+    });
+
+    templateSharedConfig.environment = Hoek.merge(templateSharedConfig.environment, yamlSharedConfig.environment || {});
+    templateSharedConfig.settings = Hoek.merge(templateSharedConfig.settings || {}, yamlSharedConfig.settings || {});
+
+    if (yamlSharedConfig.image || templateSharedConfig.image) {
+        templateSharedConfig.image = yamlSharedConfig.image || templateSharedConfig.image;
+    }
+
```

Comment:
It isn't easy to understand the intention of this logic. If the problem is that yamlSharedConfig.image should not overwrite templateSharedConfig.image when templateSharedConfig.image is truthy, then the code can be rewritten as follows:
`
if (yamlSharedConfig.image && !templateSharedConfig.image) {
    templateSharedConfig.image = yamlSharedConfig.image;
}
`